src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -196,6 +198,79 @@ public void GenerateHeader()
             PopBlock();
         }
 
+        private struct DelegateDef
+        {
+            public int numParams;
+            public List<Type> param;
+            public string pstrs, retType;
+            public string name;/*
+            DelegateDef(int numpar)
+            {
+                numParams = numpar;
+                param = new List<Type>();
+                pstrs = "";
+                retType = "";
+                name = "";
+            }*/
+        }
+
+        private static List<DelegateDef> AllDelegatesUsed = new List<DelegateDef>();
+
+        private string GenerateDelegate(Method func)
+        {
+            List<Type> allps = new List<Type>();
+            allps.Add(func.ReturnType.Type);
+            for(int i = 0; i < func.Parameters.Count; ++i)
+            {
+                allps.Add(func.Parameters.ElementAt(i).Type);
+            }
+            DelegateDef dd = new DelegateDef();
+            for(int i = 0; i < AllDelegatesUsed.Count; ++i)
+            {
+                dd = AllDelegatesUsed.ElementAt(i);
+                if (dd.numParams == func.Parameters.Count && dd.param.Equals(allps))
+                    return "DelegatesStorage." + dd.name;
+            }
+            dd.numParams = func.Parameters.Count;
+            dd.param = new List<Type>(allps);
+            dd.name = "DelegateTemplate"+ AllDelegatesUsed.Count;
+            AllDelegatesUsed.Add(dd);
+            CSharpTypePrinterResult retType;
+            var temp = GatherInternalParams(func, out retType);
+            dd.pstrs = string.Join(", ", temp);
+            dd.retType = retType.ToString();
+            return  "DelegatesStorage." + dd.name;
+        }
+
+        private void GenerateDelegatesClass()
+        {
+            if (AllDelegatesUsed.Count == 0)
+                return;
+            WriteLine("class DelegatesStorage");
+            WriteStartBraceIndent();
+            DelegateDef dd;
+            Type t;
+            string retT, pars;
+            for (int i = 0; i < AllDelegatesUsed.Count; ++i)
+            {
+                dd = AllDelegatesUsed.ElementAt(i);
+                retT = dd.param.ElementAt(0).CSharpType(TypePrinter).ToString();
+                pars = "";
+                for (int j = 1; j <= dd.numParams; ++j)
+                {
+                    t = dd.param.ElementAt(i);
```

Comment:
I am getting IndexOutOfBounds at this line. Although i am pretty sure I have done enough to ensure this wouldn't throw exception.